ocaml/xapi/xapi_pusb.ml

```diff
@@ -31,6 +32,67 @@ let create ~__context ~uSB_group ~host ~other_config ~path
   debug "PUSB ref='%s' created" (Ref.string_of pusb);
   pusb
 
+let scan_start ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let known_usb =
+    Db.PUSB.get_all_records ~__context
+    |> List.filter (fun (rf, rc) -> rc.API.pUSB_host = host)
+    |> get_known_usb
+  in
+  let local_usb =
+    get_usbs ~__context
+    |> get_local_usb
+  in
+  (* Create the newly added pusbs *)
+  USBSet.iter (fun s -> let self = create ~__context ~uSB_group:(Ref.null) ~host ~other_config:[] ~path:s.USB.path ~vendor_id:s.USB.vendor_id
+                   ~vendor_desc:s.USB.vendor_desc ~product_id:s.USB.product_id ~product_desc:s.USB.product_desc ~serial:s.USB.serial
+                   ~version:s.USB.version ~description:s.USB.description in
+               let group = Xapi_pusb_helpers.find_or_create ~__context self in
+               Db.PUSB.set_USB_group ~__context ~self ~value:group
+  ) (USBSet.diff local_usb known_usb);
+
+  List.filter (fun (rf, rc) -> USBSet.mem (extract_known_usb_info rc) (USBSet.diff known_usb local_usb)) (Db.PUSB.get_all_records ~__context)
+  |> List.iter (fun (self, _) ->
+                  let usb_group = Db.PUSB.get_USB_group ~__context ~self in
+                  Db.PUSB.destroy ~__context ~self;
+                  let vusbs = Db.USB_group.get_VUSBs ~__context ~self:usb_group in
+                  List.iter (fun vusb -> Db.VUSB.destroy ~__context ~self:vusb) vusbs;
+                  Db.USB_group.destroy ~__context ~self:usb_group)
+
+let thread_started : bool ref = ref false
+let task_processed : bool ref = ref false
 
 let scan ~__context =
-  debug "pusb scan to do"
+  (* This is for pusb-scan call which uevent will call or xapi restart. *)
+  let should_start_thread =
+    Mutex.execute mutex (
+      fun ()->
+        task_processed := false;
+        match !thread_started with
+        | false -> thread_started := true; true
+        | true -> false
+    ) in
+  (* If should_start_thread is true, then will create a thread to scan. *)
+  if should_start_thread then
+    ignore (Thread.create (
+              fun () ->
+                let rec run task_exist =
+                  match task_exist with
+                  | false -> ()
+                  | true ->
+                      let has_task = Mutex.execute mutex (
+                        fun () ->
+                          match !task_processed with
+                          | false -> task_processed := true; true
+                          | true  -> thread_started := false; false
+                       )
+                       in
+                       (* If has_task is true, then do actual start. *)
+                       if has_task then begin
+                         debug "start to scan";
```

Comment:
This needs more details as otherwise this makes no sense in the context of a log file. Consider using `__LOC__`

Reply:
yes, here this is just a reminder. I think I would remove it.